katharsis-client/src/main/java/io/katharsis/client/internal/ClientDocumentMapper.java

```diff
@@ -0,0 +1,123 @@
+
+	public ClientDocumentMapper(ResourceRegistry resourceRegistry, ObjectMapper objectMapper, PropertiesProvider propertiesProvider) {
+		super(resourceRegistry, objectMapper, propertiesProvider, true);
+		this.resourceRegistry = resourceRegistry;
+		this.objectMapper = objectMapper;
+	}
+
+	@Override
+	protected ResourceMapper newResourceMapper(final DocumentMapperUtil util, boolean client, ObjectMapper objectMapper) {
+		return new ResourceMapper(util, client, objectMapper) {
+
+			@Override
+			protected void setRelationship(Resource resource, ResourceField field, Object entity, ResourceInformation resourceInformation, QueryAdapter queryAdapter) {
+				// we also include relationship data if it is not null and not a
+				// unloaded proxy
+				boolean includeRelation = true;
+				Object relationshipValue = PropertyUtils.getProperty(entity, field.getUnderlyingName());
+				if (relationshipValue instanceof ObjectProxy) {
+					includeRelation = ((ObjectProxy) relationshipValue).isLoaded();
+				} else {
+					// TODO for fieldSets handling in the future the lazy handling must be different
+					includeRelation = relationshipValue != null || !field.isLazy() && !field.isCollection();
+				}
+
+				if(includeRelation){
+					Relationship relationship = new Relationship();
+					if(relationshipValue instanceof Collection){
+						relationship.setData(Nullable.of((Object)util.toResourceIds((Collection<?>)relationshipValue)));
+					}else{
+						relationship.setData(Nullable.of((Object) util.toResourceId(relationshipValue)));
+					}
+					resource.getRelationships().put(field.getJsonName(), relationship);		
+				}
+			}
+		};
+	}
+
+	public void setProxyFactory(ClientProxyFactory proxyFactory) {
+		this.proxyFactory = proxyFactory;
+	}
+
+	public Object fromDocument(Document document, boolean getList) {
+		ClientResourceUpsert upsert = new ClientResourceUpsert(resourceRegistry, typeParser, objectMapper, null, proxyFactory);
+
+		if (document.getErrors() != null && !document.getErrors().isEmpty()) {
+
+		}
+
+		if (!document.getData().isPresent()) {
+			return null;
+		}
+
+		List<Resource> included = document.getIncluded();
+		List<Resource> data = document.getCollectionData().get();
+
+		List<Object> dataObjects = upsert.allocateResources(data);
+		if (included != null) {
+			upsert.allocateResources(included);
+		}
+
+		upsert.setRelations(data);
+		if (included != null) {
+			upsert.setRelations(included);
+		}
+
+		if (getList) {
+			DefaultResourceList<Object> resourceList = new DefaultResourceList();
+			resourceList.addAll(dataObjects);
+			if (document.getLinks() != null) {
+				resourceList.setLinks(new JsonLinksInformation(document.getLinks(), objectMapper));
+			}
+			if (document.getMeta() != null) {
+				resourceList.setMeta(new JsonMetaInformation(document.getMeta(), objectMapper));
+			}
+			return resourceList;
+		} else {
+			if(dataObjects.isEmpty()){
+				return null;
+			}
+			if(dataObjects.size() != 1){
```

Comment:
this line should be moved up and just be `if(dataObjets.size > 1)` because it's contradictory to line 113